File: services/matchmaking.js
const db = require('../database');

/**
 * Matchmaking Service: Manages Ranked (ELO) and Casual (TPS) queues.
 */
let queue = []; 

module.exports = {
    // Callback to be set by the battle system
    onMatchFound: null,

    /**
     * TPS = sum((rarityWeight * level)) + (total_upgrades * 10)
     */
    async calculateTPS(user) {
        const roster = await db.roster.find({ userId: user.telegramId });
        let tps = 0;
        const weights = { "Common": 1, "Rare": 2, "Epic": 3, "Legendary": 4 };

        for (const charId of (user.teamIds || [])) {
            if (!charId) continue;
            const entry = roster.find(r => r.charId === charId);
            if (entry) {
                const weight = weights[entry.rarity] || 1;
                const upgradesCount = Object.keys(entry.upgrades || {}).length;
                tps += (weight * entry.level) + (upgradesCount * 10);
            }
        }
        return tps;
    },

    async joinQueue(user, isCasual = false) {
        // 1. Basic Checks
        const now = Date.now();
        const lastAction = user.lastQueueJoin || 0;
        if (now - lastAction < 10000) {
             return { error: `⏳ Please wait ${Math.ceil((10000 - (now - lastAction))/1000)}s before queuing again.` };
        }

        if (!user.teamIds || user.teamIds.filter(id => id).length < 3) {
            return { error: "❌ Your team must have 3 characters. Use /team to set your lineup." };
        }

        // 2. Prepare Entry
        this.leaveQueue(user.telegramId);
        const tps = await this.calculateTPS(user);
        
        const entry = {
            userId: user.telegramId,
            username: user.username,
            elo: user.elo || 1000,
            tps: tps,
            isCasual: isCasual,
            timestamp: now
        };

        await db.users.update({ telegramId: user.telegramId }, { $set: { lastQueueJoin: now } });
        queue.push(entry);
        return { success: true, tps };
    },

    leaveQueue(userId) {
        queue = queue.filter(q => q.userId !== userId);
    },

    /**
     * Matches players in the queue. Runs via Interval.
     */
    async processQueue() {
        const now = Date.now();
        // Timeouts (60s)
        const active = queue.filter(q => (now - q.timestamp) < 60000);
        const timedOut = queue.filter(q => (now - q.timestamp) >= 60000);
        
        // Notify timed out
        for (const q of timedOut) {
            this.leaveQueue(q.userId);
            if (this.onMatchTimeout) this.onMatchTimeout(q.userId);
        }

        queue = active;

        const matched = new Set();

        for (let i = 0; i < queue.length; i++) {
            const p1 = queue[i];
            if (matched.has(p1.userId)) continue;

            for (let j = i + 1; j < queue.length; j++) {
                const p2 = queue[j];
                if (matched.has(p2.userId)) continue;

                if (p1.isCasual !== p2.isCasual) continue;

                let isMatch = false;
                if (p1.isCasual) {
                    isMatch = Math.abs(p1.tps - p2.tps) <= 200;
                } else {
                    isMatch = Math.abs(p1.elo - p2.elo) <= 100;
                }

                if (isMatch) {
                    matched.add(p1.userId);
                    matched.add(p2.userId);
                    this.leaveQueue(p1.userId);
                    this.leaveQueue(p2.userId);

                    if (this.onMatchFound) {
                        this.onMatchFound(p1, p2, p1.isCasual ? 'casual' : 'ranked');
                    }
                    break;
                }
            }
        }
    }
};

// Start Matchmaking Cycle
setInterval(() => module.exports.processQueue(), 3000);

File: services/matchmaking.py
import time
import asyncio
from database import db

class MatchmakingService:
    def __init__(self):
        self.queue = []
        self.on_match_found = None
        self.on_match_timeout = None

    async def calculate_tps(self, user):
        roster = await db.roster.find({"userId": user['telegramId']})
        tps = 0
        weights = {"Common": 1, "Rare": 2, "Epic": 3, "Legendary": 4, "Mythic": 5}
        
        team_ids = user.get('teamIds', [])
        for char_id in team_ids:
            if not char_id: continue
            entry = next((r for r in roster if r['charId'] == char_id), None)
            if entry:
                weight = weights.get(entry.get('rarity', 'Common'), 1)
                upgrades_count = len(entry.get('upgrades', {}))
                tps += (weight * entry.get('level', 1)) + (upgrades_count * 10)
        return tps

    async def join_queue(self, user, is_casual=False):
        now = int(time.time() * 1000)
        last_action = user.get('lastQueueJoin', 0)
        
        if now - last_action < 10000:
            wait = int((10000 - (now - last_action)) / 1000)
            return {"error": f"⏳ Please wait {wait}s before queuing again."}

        team_ids = user.get('teamIds', [])
        if len([tid for tid in team_ids if tid]) < 3:
            return {"error": "❌ Your team must have 3 characters. Use /myscorer to set your lineup."}

        self.leave_queue(user['telegramId'])
        tps = await self.calculate_tps(user)
        
        entry = {
            "userId": user['telegramId'],
            "username": user.get('username', 'Sorcerer'),
            "elo": user.get('elo', 1000),
            "tps": tps,
            "isCasual": is_casual,
            "timestamp": now
        }

        await db.users.update({"telegramId": user['telegramId']}, {"$set": {"lastQueueJoin": now}})
        self.queue.append(entry)
        return {"success": True, "tps": tps}

    def leave_queue(self, user_id):
        self.queue = [q for q in self.queue if q['userId'] != user_id]

    async def process_queue(self):
        while True:
            await asyncio.sleep(3)
            now = int(time.time() * 1000)
            
            # Timeouts (60s)
            active = [q for q in self.queue if (now - q['timestamp']) < 60000]
            timed_out = [q for q in self.queue if (now - q['timestamp']) >= 60000]
            
            for q in timed_out:
                self.leave_queue(q['userId'])
                if self.on_match_timeout:
                    await self.on_match_timeout(q['userId'])

            self.queue = active
            matched_ids = set()

            for i in range(len(self.queue)):
                p1 = self.queue[i]
                if p1['userId'] in matched_ids: continue

                for j in range(i + 1, len(self.queue)):
                    p2 = self.queue[j]
                    if p2['userId'] in matched_ids: continue
                    if p1['isCasual'] != p2['isCasual']: continue

                    is_match = False
                    if p1['isCasual']:
                        is_match = abs(p1['tps'] - p2['tps']) <= 200
                    else:
                        is_match = abs(p1['elo'] - p2['elo']) <= 100

                    if is_match:
                        matched_ids.add(p1['userId'])
                        matched_ids.add(p2['userId'])
                        
                        if self.on_match_found:
                            mode = 'casual' if p1['isCasual'] else 'ranked'
                            # In Python, we might need a reference to the bot or dispatcher
                            asyncio.create_task(self.on_match_found(p1, p2, mode))
                        
                        break
            
            # Clean up matched players
            self.queue = [q for q in self.queue if q['userId'] not in matched_ids]

matchmaking_service = MatchmakingService()
